lib/statsd/instrument/batched_udp_sink.rb

```diff
@@ -0,0 +1,124 @@
+# frozen_string_literal: true
+
+module StatsD
+  module Instrument
+    # @note This class is part of the new Client implementation that is intended
+    #   to become the new default in the next major release of this library.
+    class BatchedUDPSink
+      DEFAULT_FLUSH_INTERVAL = 1.0
+      MAX_PACKET_SIZE = 508
+
+      def self.for_addr(addr, flush_interval: DEFAULT_FLUSH_INTERVAL)
+        host, port_as_string = addr.split(":", 2)
+        new(host, Integer(port_as_string), flush_interval: flush_interval)
+      end
+
+      attr_reader :host, :port
+
+      def initialize(host, port, flush_interval: DEFAULT_FLUSH_INTERVAL)
+        @host = host
+        @port = port
+        @mutex = Mutex.new
+        @socket = nil
+        @flush_interval = flush_interval
+
+        require "concurrent/array"
+        @buffer = Concurrent::Array.new
```

Comment:
What is difference with the built-in Queue class? Less overhead?

Reply:
No, it's just a very different usage. `Queue` allows you to block until you get an element. But it doesn't have any method allowing you to pop more than one element, so it's not suitable for batching.

`Concurrent::Array` is basically MRI's regular array, but thread safe. On MRI it's just `::Array`, for other implementations there some synchronization.

Reply:
An alternative would be to use a String buffer, which we could even use to build the datagram into directly and could be flushed when it reaches its capacity.  Even without batching, that could provide some benefits of reducing string allocations and resizing during datagram building.  Although, that datagram building refactoring is probably out of scope for this PR.

Besides, we may want to decouple the concept of a buffered sink from the concerns sink sharing (e.g. per-thread buffers are an alternative) and periodic flushing (e.g. block based buffering is an alternative)